extmod/ssl/ssl.py

```diff
@@ -0,0 +1,172 @@
+        #
+
+    @property
+    def protocol(self):
+        return self._protocol
+
+    @property
+    def check_hostname(self):
+        return self._check_hostname
+
+    @check_hostname.setter
+    def check_hostname(self, value):
+        assert type(value) == bool
+
+        if value is True and not self.verify_mode:
+            self.verify_mode = CERT_REQUIRED
+
+        self._check_hostname = value
+
+    @property
+    def verify_mode(self):
+        return self._verify_mode
+
+    @verify_mode.setter
+    def verify_mode(self, value):
+        assert value in (CERT_NONE, CERT_OPTIONAL, CERT_REQUIRED)
+        if not self.check_hostname:
+            self._verify_mode = value
+
+    def _reset(self, server=True):
+        ctx = _ussl.ctx_init()
+        if server:
+            ctx.load_certchain(key=self.key, cert=self.cert)
+            if self.cadata:
+                ctx.load_cadata(self.cadata)
+        else:
+            if self.cadata:
+                ctx.load_cadata(self.cadata)
+            if self.key:
+                ctx.load_certchain(key=self.key, cert=self.cert)
+
+        if self.ciphersuite:
+            ctx.set_ciphers(self.ciphersuite)
+        return ctx
+
+    def load_cert_chain(self, certfile, keyfile=None):
+        # load certificate/key from a file, is possible to use memoryview to
+        # save RAM?
+        if isinstance(certfile, bytes):
+            self.cert = certfile
+        else:
+            with open(certfile, "rb") as cert:
+                self.cert = cert.read()
+        if keyfile:
+            if isinstance(keyfile, bytes):
+                self.key = keyfile
+            else:
+                with open(keyfile, "rb") as key:
+                    self.key = key.read()
+
+        self.ctx.load_certchain(key=self.key, cert=self.cert)
+
+    def load_default_certs(self):
+        pass  # not sure how/if implement this
+
+    def load_verify_locations(self, cafile=None, capath=None, cadata=None):
+        if cafile:
+            with open(cafile, "rb") as ca_cert:
+                self.cadata = ca_cert.read()
+        if capath:
+            pass  ## not implemented, it needs to concatenate multiple files
+
+        if cadata:
+            self.cadata = cadata
+
+        self.ctx.load_cadata(self.cadata)
+
+    def get_ciphers(self):
+        return self.ctx.get_ciphers()
+
+    def set_ciphers(self, ciphersuite):
+        self.ciphersuite = ciphersuite.split(":")
+        return self.ctx.set_ciphers(self.ciphersuite)
+
+    def wrap_socket(
+        self,
+        sock,
+        server_side=False,
+        do_handshake_on_connect=True,
+        server_hostname=None,
+    ):
+        if self.check_hostname and server_hostname is None:
+            raise ValueError("check_hostname requires server_hostname")
+        # to be substituted by e.g. _ussl._context_wrap_socket in modussl_mbedtls.c ?:
+        # _ussl._context_wrap_socket(*args, **kargs)
+        if self._reload_ctx:
```

Comment:
I think this `_reload_ctx` logic can be removed if the mbedtls state is properly split between SSLContext and SSLSocket (as discussed above).